README.md
[![Chat on Gitter](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/genotrance/px)
[![Chat on Matrix](https://img.shields.io/matrix/genotrance_px:matrix.org)](https://matrix.to/#/#genotrance_px:matrix.org)

# Px

## What is Px?
Px is a HTTP(s) proxy server that allows applications to authenticate through
an NTLM or Kerberos proxy server, typically used in corporate deployments,
without having to deal with the actual handshake. It is primarily designed to
run on Windows systems and automatically authenticates on using the currently
logged in Windows user account. It is also possible to run Px on Windows, Linux
and MacOS when single signon is not available by configuring the domain, username
and password to authenticate with.

Px is similar to "NTLM Authorization Proxy Server" [NTLMAPS](http://ntlmaps.sourceforge.net/)
and [Cntlm](http://cntlm.sourceforge.net/) in that it sits between the corporate
proxy and applications and offloads authentication. The advantage is that Px is
able to use the currently logged in user's credentials automatically without
requiring any user supplied credentials. This is accomplished by using Microsoft
SSPI to generate the tokens and signatures required to authenticate with the proxy.

Px also supports Kerberos and works with user supplied credentials for cases
where SSPI is not available. Kerberos/Negotiate support is only available on
Windows at this time.

Microsoft provides a good starting point to understand how NTLM [authentication](https://msdn.microsoft.com/en-us/library/dd925287.aspx)
works. And similarly for [Kerberos](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc772815(v=ws.10)) (warning: long!)

## Installation

If Python is already available, Px can be easily installed using the Python
package manager `pip`. This will download and install Px as a Python module
along with all dependencies:

- Latest:

	`pip install git+https://github.com/genotrance/px`

- Stable:

	`pip install px-proxy`

The latest Px version can also be downloaded and installed from source:

- Via git:

	`git clone https://github.com/genotrance/px`

- Download [ZIP](https://github.com/genotrance/px/archive/master.zip):

	`https://github.com/genotrance/px/archive/master.zip`

Once downloaded, Px can be installed as a standard Python module along
with all dependencies :

  `python setup.py install`

Once installed, Px can be run as follows:
- In the background: `pythonw -m px`
- In the foreground in a console window: `python -m px`

Lastly, Px can be run as a local Python script without installation. Download the source
as described above, install all dependencies manually using `pip` and then run Px:

```
pip install keyring netaddr ntlm-auth psutil pywin32 winkerberos futures

pythonw px.py # run in the background
python px.py # run in a console window
```

If installed, Px can be uninstalled as follows:

```
python -m px --uninstall
pip uninstall px-proxy
```

If Python is not available, download the latest Windows binary ZIP from the [releases](https://github.com/genotrance/px/releases)
page. This binary is compiled using [Nuitka](https://nuitka.net) and contains everything
needed to run standalone.

## Configuration

Px requires only one piece of information in order to function - the server
name and port of the proxy server. This needs to be configured in px.ini. If not
specified, Px will check Internet Options or environment variables for any proxy
definitions and use them. Without this, Px will try to connect to sites directly.

The noproxy capability allows Px to connect to hosts in the configured subnets
directly, bypassing the proxy altogether. This allows clients to connect to
hosts within the intranet without requiring additional configuration for each
client or at the proxy.

### Credentials

If SSPI is not available or not preferred, providing a `username` in `domain\username`
format allows Px to authenticate as that user. The corresponding password is
retrieved using Python keyring and needs to be setup directly in the backend.

On Windows, Credential Manager is the backend and can be accessed as follows:

    Control Panel > User Accounts > Credential Manager > Windows Credentials

Or run command line

    rundll32.exe keymgr.dll, KRShowKeyMgr

On Linux, the Gnome Keyring or KWallet can be used to store passwords.

Px looks for a 'generic credential' type with 'Px' as the network address name. More
information on keyring backends can be found [here](https://pypi.org/project/keyring).

### Misc

There are a few other settings to tweak in the INI file but most are obvious.
All settings can be specified on the command line for convenience. The INI file
can also be created or updated from the command line using `--save`.

The binary distribution of Px runs in the background once started and can be
quit by running `px --quit`. When run directly using Python, use `CTRL-C` to quit.

Px can also be setup to automatically run on startup with the --install flag.
This is done by adding an entry into the Window registry which can be removed
with `--uninstall`.

NOTE: Command line parameters passed with `--install` are not saved for use on
startup. The `--save` flag or manual editing of `px.ini` is required to provide
configuration to Px on startup.

## Usage

```
px [FLAGS]
python px.py [FLAGS]

Actions:
  --save
  Save configuration to px.ini or file specified with --config
    Allows setting up Px config directly from command line
    Values specified on CLI override any values in existing config file
    Values not specified on CLI or config file are set to defaults

  --install
  Add Px to the Windows registry to run on startup

  --uninstall
  Remove Px from the Windows registry

  --quit
  Quit a running instance of Px.exe

Configuration:
  --config=
  Specify config file. Valid file path, default: px.ini in working directory

  --proxy=  --server=  proxy:server= in INI file
  NTLM server(s) to connect through. IP:port, hostname:port
    Multiple proxies can be specified comma separated. Px will iterate through
    and use the one that works

  --pac=  proxy:pac=
  PAC file to use to connect
    Use in place of server if PAC file should be loaded from a custom URL or
    file location instead of from Internet Options

  --listen=  proxy:listen=
  IP interface to listen on. Valid IP address, default: 127.0.0.1

  --port=  proxy:port=
  Port to run this proxy. Valid port number, default: 3128

  --gateway  proxy:gateway=
  Allow remote machines to use proxy. 0 or 1, default: 0
    Overrides 'listen' and binds to all interfaces

  --hostonly  proxy:hostonly=
  Allow only local interfaces to use proxy. 0 or 1, default: 0
    Px allows all IP addresses assigned to local interfaces to use the service.
    This allows local apps as well as VM or container apps to use Px when in a
    NAT config. Px does this by listening on all interfaces and overriding the
    allow list.

  --allow=  proxy:allow=
  Allow connection from specific subnets. Comma separated, default: *.*.*.*
    Whitelist which IPs can use the proxy. --hostonly overrides any definitions
    unless --gateway mode is also specified
    127.0.0.1 - specific ip
    192.168.0.* - wildcards
    192.168.0.1-192.168.0.255 - ranges
    192.168.0.1/24 - CIDR

  --noproxy=  proxy:noproxy=
  Direct connect to specific subnets like a regular proxy. Comma separated
    Skip the NTLM proxy for connections to these subnets
    127.0.0.1 - specific ip
    192.168.0.* - wildcards
    192.168.0.1-192.168.0.255 - ranges
    192.168.0.1/24 - CIDR

  --useragent=  proxy:useragent=
  Override or send User-Agent header on client's behalf

  --username=  proxy:username=
  Authentication to use when SSPI is unavailable. Format is domain\\username
  Service name "Px" and this username are used to retrieve the password using
  Python keyring. Px only retrieves credentials and storage should be done
  directly in the keyring backend.
    On Windows, Credential Manager is the backed and can be accessed from
    Control Panel > User Accounts > Credential Manager > Windows Credentials.
    Create a generic credential with Px as the network address, this username
    and corresponding password.

  --auth=  proxy:auth=
  Force instead of discovering upstream proxy type
    By default, Px will attempt to discover the upstream proxy type and either
    use pywin32/ntlm-auth for NTLM auth or winkerberos for Kerberos or Negotiate
    auth. This option will force either NTLM, Kerberos or Basic and not query the
    upstream proxy type.

  --workers=  settings:workers=
  Number of parallel workers (processes). Valid integer, default: 2

  --threads=  settings:threads=
  Number of parallel threads per worker (process). Valid integer, default: 5

  --idle=  settings:idle=
  Idle timeout in seconds for HTTP connect sessions. Valid integer, default: 30

  --socktimeout=  settings:socktimeout=
  Timeout in seconds for connections before giving up. Valid float, default: 20

  --proxyreload=  settings:proxyreload=
  Time interval in seconds before refreshing proxy info. Valid int, default: 60
    Proxy info reloaded from manual proxy info defined in Internet Options

  --foreground  settings:foreground=
  Run in foreground when frozen or with pythonw.exe. 0 or 1, default: 0
    Px will attach to the console and write to it even though the prompt is
    available for further commands. CTRL-C in the console will exit Px

  --debug  settings:log=
  Enable debug logging. default: 0
    Logs are written to working directory and over-written on startup
    A log is automatically created if Px crashes for some reason

  --uniqlog
  Generate unique log file names
    Prevents logs from being overwritten on subsequent runs. Also useful if
    running multiple instances of Px
```

## Examples

Use `proxyserver.com:80` and allow requests from localhost only:

  `px --proxy=proxyserver.com:80`

Don't use any forward proxy at all, just log what's going on:

  `px --noproxy=0.0.0.0/0 --debug`

Allow requests from `localhost` and all locally assigned IP addresses. This
is very useful for Docker for Windows and VMs in a NAT configuration because
all requests originate from the host's IP:

  `px --proxy=proxyserver.com:80 --hostonly`

Allow requests from `localhost`, locally assigned IP addresses and the IPs
specified in the allow list outside the host:

  `px --proxy=proxyserver:80 --hostonly --gateway --allow=172.*.*.*`

Allow requests from everywhere. Be careful, every client will use your login:

  `px --proxy=proxyserver.com:80 --gateway`

NOTE: In Docker for Windows you need to set your proxy to `http://host.docker.internal:3128` or `http://<your_ip>:3128`
(or actual port Px is listening to) in your containers and be aware of https://github.com/docker/for-win/issues/1380.

Workaround:

`docker build --build-arg http_proxy=http://<your ip>:3128 --build-arg https_proxy=http://<your ip>:3128 -t containername ../dir/with/Dockerfile`

## Dependencies

Px doesn't have any GUI and runs completely in the background. It depends on
the following Python packages:

  `keyring`, `netaddr`, `ntlm-auth`, `psutil`

  `pywin32`, `winkerberos` on Windows

  `futures` on Python 2.x

In order to make Px a capable proxy server, it is designed to run in multiple
processes. The number of parallel workers or processes is configurable. However,
this only works on Python 3.3+ since that's when support was added to share
sockets across processes in Windows. On older versions of Python, Px will run
multi-threaded but in a single process. The number of threads per process is
also configurable.

## Building

To build a self-sufficient executable that does not depend on the presence of
Python and dependency modules, both Nuitka and PyInstaller scripts are provided.
Run `python tools.py` for more details.

## Feedback

Px is definitely a work in progress and any feedback or suggestions are welcome.
It is hosted on [GitHub](https://github.com/genotrance/px) with an MIT license
so issues, forks and PRs are most appreciated. Join us on the [discussion](https://github.com/genotrance/px/discussions)
board, [Gitter](https://gitter.im/genotrance/px) or [Matrix](https://matrix.to/#/#genotrance_px:matrix.org)
to chat about Px.

## Credits

Thank you to all [contributors](https://github.com/genotrance/px/graphs/contributors) for their PRs and all issue submitters.


HISTORY.txt
v0.6.0 - TBD
- Moved all Windows proxy detection code into wproxy.py, eventually to be made
  an independent module
- Moved debugging code into separate debug.py module
- Removed support for file:// PAC files configured in network settings since
  Windows 10 no longer supports this scheme (see https://bit.ly/3JKPgjR)
- Added support in wproxy to detect proxies defined via environment variables
- Added support for Linux - only NTLM and BASIC authentication are supported
  for now

v0.5.1 - 2022-03-22
- Fixed #128 - IP:port split once from the right
- Binary is now built using Nuitka
- Moved PyInstaller build script to tools.py
- Removed px.ico from PyInstaller build since it upsets anti-virus tools

v0.5.0 - 2022-01-26
- Added support for issue #58 - authentication with user/password when SSPI is
  unavailable or not preferred.
- Pulled PR #68 to handle case where no longer behind PAC proxy
- Fixed minor bug in file:// PAC file handling
- Implemented #65 - support for specifying PAC in INI
- Implemented #73 - force auth mechanism in INI
- Merged PR #85 - sanitize authorization header, shutdown socket, defer
  connection header during auth (#71), drop completely for CONNECT, forward
  error body, 2.7 fix
- Switched to Python 3.7, drop 3.4 support
- Merged PR #80 - fallback to DIRECT if no proxy found via auto or PAC discovery
- Added basic auth support (PR #82)
- Fixed issue where an empty domain was breaking password access
- Fixed #88, issue 2 in #71, PR #83 - use auth mechanism in order
- Added code to keep-alive during authentication - issue 3 in #71
- Fixed #108 - case sensitive authentication response
- Fixed #116 - Python 3.0 deprecation compatibility
- Fixed #122 - pywin32 import order
- Resolved #72 via documentation update
- Minor changes to improve client and proxy disconnects

v0.4.0 - 2018-09-04
- Implemented #18 - support for multiple NTLM proxies that will be iterated
  through if connection fails
- Included fix for allowing parent proxy to resolve hostname if not resolvable
  locally - #18
- Added --socktimeout to configure maximum time to wait before giving up on
  socket connections
- Fixed #27 - Px does not exit with --quit
- Fixed #26 - Px doesn't attach to powershell console
- Allow Px to work with pythonw.exe - console output and stdout crash
- Fix Python 2.x crashes in quit()
- Fixed issue #22 - support for Kerberos authentication
- Removed pywin32 option since it doesn't work for Kerberos auth
- Fixed issues with useragent header override functionality
- Allow Px to run in the foreground when frozen or run with pythonw.exe - #25
- Make --proxy an optional setting if --noproxy is defined, reject connections
  if remote server isn't in noproxy list and no proxy defined
- Add --hostonly which only allows IP addresses assigned to local interfaces to
  use Px. Disallow any external clients unless --gateway specified along with a
  restrictive --allow definition - PR20
- Changed --socktimeout from int to float
- Added support to discover proxy info from Internet Options - Automatic Config
  URL with PAC files, WPAD, or static proxy definition - #30
- Added --proxyreload flag to configure interval between rediscovery of proxy
  info
- Merged PR32 to fix PAC proxy port detection
- Implemented fix for #29 - discover proxy type instead of assuming Negotiate
  will work across the board
- Close proxy socket if it returns "Proxy-Connection: close" header
- Improved memory and disk usage which had increased drastically after adding
  PAC support - #30
- Added setup.py to install Px using pip - #24
- Fixed bug where Px crashes on Windows startup - #31
- Fixed issue of no debug output - #36
- Ensure logs are generated where Px is instead of working directory and flushed
  immediately, quit should log in separate file and not overwrite MainProcess
- Added --uniqlog to allow multiple instances of Px to run with separate log
  destinations or to prevent overwriting logs on subsequent runs
- Sanitize auth exchanges in generated logs for easy sharing in issues
- Handle NTLM auth failure instead of crashing - #34
- Add FTP port 21 for ftp:// requests - #39
- Catch exception for issue #38
- Fix issue #44 - move from README.txt to README.md
- Fix issue #43 - fix PAC support using WINHTTP
- Added documentation for #46
- Include PR #48 to enable support for failover proxies received from PAC
- Fix issue #51 - support for Windows 8 and below
- Fix issue #34 by using pywin32 for NTLM and winkerberos for Kerberos since the
  latter doesn't work consistently for NTLM
- Pulled threading improvements from PR #56
- Improve message for issue #5
- Fix issue #57 - handle pip shim scenario during --install
- Fix issue #47 and #52 by forwarding data as received
- Fix issue #60 - chunked encoding transfer broken
- Increasing default socktimeout to 20 seconds

v0.3.0 - 2018-02-19
- Fixed issue 9 - Added support for winkerberos to workaround pywin32 bug
- Cleaned up loading module dependencies across Python versions
- Fixed issue where thread name in logs was wrong due to Python 3.6+ issue
- Fixed issue where URL without trailing / was sending blank path to remote
- Fixed issue where Px was disconnecting instead of keeping connections alive
- Changed Pyinstaller to build a single executable file
- Fixed bug where chunked transfer headers were being returned incorrectly
- Removed sleep from non-CONNECT methods that was slowing down Px considerably
- Added "allow" feature to allow (deny) client ip addresses (helpful with
  Docker for Windows)
- Added "--gateway" parameter
- Fixed issue where a HTTP 304 Not Modified response would hang
- Enabled --debug also for --quit code path
- Fixed issue where --quit would not be able to stop px instances with
  different uppercase/lowercase process names
- Fixed exception when logging bad requests containing unicode
- Fixed issue 12 - proxy keep-alive support for HTTP/1.0 clients
- Fixed issue 13 - support for HTTP Expect-Continue

wproxy.py
"Load proxy information from the operating system"

import copy
import socket
import sys

# Python 2.x vs 3.x support
try:
    import urllib.parse as urlparse
    import urllib.request as request
except ImportError:
    import urlparse
    import urllib as request

try:
    import netaddr
except ImportError:
    pprint("Requires module netaddr")
    sys.exit()

# Proxy modes - source of proxy info
MODE_NONE = 0
MODE_AUTO = 1
MODE_PAC = 2
MODE_MANUAL = 3
MODE_ENV = 4
MODE_CONFIG = 5
MODE_CONFIG_PAC = 6

# Direct proxy connection
DIRECT = ("DIRECT", 80)

# Debug shortcut
dprint = lambda x: None

def parse_proxy(proxystrs):
    """
    Convert comma separated list of proxy:port into list[tuple(host,port)]
      If no port, default to 80
      Raises ValueError if bad proxy format
    """

    if not proxystrs or len(proxystrs) == 0:
        return []

    servers = []
    for proxystr in [i.strip() for i in proxystrs.split(",")]:
        pserver = [i.strip() for i in proxystr.rsplit(":", 1)]
        if len(pserver) == 1:
            pserver.append(80)
        elif len(pserver) == 2:
            try:
                pserver[1] = int(pserver[1])
            except ValueError as error:
                raise ValueError("Bad proxy server port: " + pserver[1]) from error
        else:
            raise ValueError("Bad proxy server definition: " + proxystr)

        if tuple(pserver) not in servers:
            servers.append(tuple(pserver))

    return servers

def parse_noproxy(noproxystr, iponly = False):
    """
    Convert comma/semicolon separated noproxy list of IP1,IP2,host1,host2 and returns two
    lists: one with IP addreses and ranges, second with hostnames
      If iponly = True, only allows returns IP addresses and ranges
    """

    noproxy = netaddr.IPSet([])
    noproxy_hosts = []

    bypasses = [h for h in noproxystr.lower().replace(' ', ',').replace(';', ',').split(',')]
    for bypass in bypasses:
        if len(bypass) == 0:
            continue

        try:
            if "-" in bypass:
                spl = bypass.split("-", 1)
                ipns = netaddr.IPRange(spl[0], spl[1])
            elif "*" in bypass:
                ipns = netaddr.IPGlob(bypass)
            else:
                ipns = netaddr.IPNetwork(bypass)
            noproxy.add(ipns)
            dprint("Noproxy += " + bypass)
        except Exception as error:
            if not iponly:
                if bypass == "<local>":
                    # TODO: detect all intranet addresses
                    noproxy_hosts.append("localhost")
                    noproxy.add(netaddr.IPNetwork("127.0.0.1/24"))
                else:
                    noproxy_hosts.append(bypass)
                dprint("Noproxy hostname += " + bypass)
            else:
                dprint("Bad IP definition: %s" % bypass)
                raise error

    if iponly:
        return noproxy
    else:
        return noproxy, noproxy_hosts

class _WproxyBase(object):
    "Load proxy information using urllib.request.getproxies()"

    mode = MODE_NONE
    servers = None
    noproxy = None
    noproxy_hosts = None

    def __init__(self, mode = MODE_NONE, servers = None, noproxy = None, noproxy_hosts = None, debug_print = None):
        global dprint
        if debug_print is not None:
            dprint = debug_print

        self.servers = []
        self.noproxy = netaddr.IPSet([])
        self.noproxy_hosts = []

        if mode != MODE_NONE:
            # MODE_CONFIG or MODE_CONFIG_PAC
            self.mode = mode
            self.servers = servers or []
            self.noproxy = noproxy or netaddr.IPSet([])
            self.noproxy_hosts = noproxy_hosts or []
        else:
            proxy = request.getproxies()
            if "http" in proxy:
                self.mode = MODE_ENV
                self.servers = parse_proxy(proxy["http"])

                if "no" in proxy:
                    self.noproxy, self.noproxy_hosts = parse_noproxy(proxy["no"])

    def get_netloc(self, url):
        "Split url into netloc = hostname:port and path"

        nloc = url
        parse = urlparse.urlparse(url, allow_fragments=False)
        if parse.netloc:
            nloc = parse.netloc
        if ":" not in nloc:
            port = parse.port
            if not port:
                if parse.scheme == "http":
                    port = 80
                elif parse.scheme == "https":
                    port = 443
                elif parse.scheme == "ftp":
                    port = 21
            netloc = (nloc, port)
        else:
            spl = nloc.rsplit(":", 1)
            netloc = (spl[0], int(spl[1]))

        path = parse.path or "/"
        if parse.params:
            path = path + ";" + parse.params
        if parse.query:
            path = path + "?" + parse.query

        dprint("netloc = %s, path = %s" % (netloc, path))

        return netloc, path

    def check_noproxy_for_netloc(self, netloc):
        """
        Check if (host, port) is in noproxy list
        Returns:
          (IP, port) of netloc to connect directly
          None to connect through proxy
        """

        if self.noproxy.size:
            addr = []
            try:
                addr = socket.getaddrinfo(netloc[0], netloc[1])
            except socket.gaierror:
                # Couldn't resolve, let parent proxy try, px#18
                dprint("Couldn't resolve host")
            if len(addr) != 0 and len(addr[0]) == 5:
                ipport = addr[0][4]
                # "%s => %s + %s" % (url, ipport, path)

                if ipport[0] in self.noproxy:
                    # Direct connection from noproxy configuration
                    return ipport

    def check_noproxy_for_url(self, url):
        """
        Check if url is in noproxy list
        Returns:
          (IP, port) of URL to connect directly
          None to connect through proxy
        """

        if self.noproxy.size:
            netloc, path = self.get_netloc(url)
            return self.check_noproxy_for_netloc(netloc)

    def find_proxy_for_url(self, url):
        """
        Return list of proxy servers to use for this url
            DIRECT can be returned as one of the options in the response
        """

        netloc, path = self.get_netloc(url)
        if self.mode == MODE_NONE:
            # Direct connection since no proxy configured
            return [DIRECT], netloc, path

        ipport = self.check_noproxy_for_netloc(netloc)
        if ipport is not None:
            # Direct connection since in noproxy list
            return [DIRECT], ipport, path

        if self.mode == MODE_ENV:
            # Return proxy from the environment
            return copy.deepcopy(self.servers), netloc, path

        return None, netloc, path

if sys.platform == "win32":
    import ctypes
    import ctypes.wintypes

    # Windows version
    #  6.1 = Windows 7
    #  6.2 = Windows 8
    #  6.3 = Windows 8.1
    # 10.0 = Windows 10
    WIN_VERSION = float(
        str(sys.getwindowsversion().major) + "." +
        str(sys.getwindowsversion().minor))

    class WINHTTP_CURRENT_USER_IE_PROXY_CONFIG(ctypes.Structure):
        _fields_ = [("fAutoDetect", ctypes.wintypes.BOOL),
                    # "Automatically detect settings"
                    ("lpszAutoConfigUrl", ctypes.wintypes.LPWSTR),
                    # "Use automatic configuration script, Address"
                    ("lpszProxy", ctypes.wintypes.LPWSTR),
                    # "1.2.3.4:5" if "Use the same proxy server for all protocols",
                    # else advanced
                    # "ftp=1.2.3.4:5;http=1.2.3.4:5;https=1.2.3.4:5;socks=1.2.3.4:5"
                    ("lpszProxyBypass", ctypes.wintypes.LPWSTR),
                    # ";"-separated list
                    # "Bypass proxy server for local addresses" adds "<local>"
                ]

    class WINHTTP_AUTOPROXY_OPTIONS(ctypes.Structure):
        _fields_ = [("dwFlags", ctypes.wintypes.DWORD),
                    ("dwAutoDetectFlags", ctypes.wintypes.DWORD),
                    ("lpszAutoConfigUrl", ctypes.wintypes.LPCWSTR),
                    ("lpvReserved", ctypes.c_void_p),
                    ("dwReserved", ctypes.wintypes.DWORD),
                    ("fAutoLogonIfChallenged", ctypes.wintypes.BOOL), ]

    class WINHTTP_PROXY_INFO(ctypes.Structure):
        _fields_ = [("dwAccessType", ctypes.wintypes.DWORD),
                    ("lpszProxy", ctypes.wintypes.LPWSTR),
                    ("lpszProxyBypass", ctypes.wintypes.LPWSTR), ]

    # Parameters for WinHttpOpen, http://msdn.microsoft.com/en-us/library/aa384098(VS.85).aspx
    WINHTTP_NO_PROXY_NAME = 0
    WINHTTP_NO_PROXY_BYPASS = 0
    WINHTTP_FLAG_ASYNC = 0x10000000

    # dwFlags values
    WINHTTP_AUTOPROXY_AUTO_DETECT = 0x00000001
    WINHTTP_AUTOPROXY_CONFIG_URL = 0x00000002

    # dwAutoDetectFlags values
    WINHTTP_AUTO_DETECT_TYPE_DHCP = 0x00000001
    WINHTTP_AUTO_DETECT_TYPE_DNS_A = 0x00000002

    # dwAccessType values
    WINHTTP_ACCESS_TYPE_DEFAULT_PROXY = 0
    WINHTTP_ACCESS_TYPE_NO_PROXY = 1
    WINHTTP_ACCESS_TYPE_NAMED_PROXY = 3
    WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY = 4

    # Error messages
    WINHTTP_ERROR_WINHTTP_UNABLE_TO_DOWNLOAD_SCRIPT = 12167
    WINHTTP_ERROR_WINHTTP_AUTODETECTION_FAILED = 12180

    class Wproxy(_WproxyBase):
        "Load proxy information from Windows Internet Options"

        def __init__(self, mode = MODE_NONE, servers = None, noproxy = None, noproxy_hosts = None, debug_print = None):
            """
            Load proxy information from Windows Internet Options
              Returns MODE_NONE, MODE_ENV, MODE_AUTO, MODE_PAC, MODE_MANUAL

            Mode can be set to MODE_CONFIG or MODE_CONFIG_PAC
              MODE_CONFIG expects servers = [(proxy1, port), (proxy2, port)]
              MODE_CONFIG_PAC expects servers = [pac_url]

            Order of proxy priority:
            - Configuration - MODE_CONFIG, MODE_CONFIG_PAC
            - Internet Options - MODE_AUTO, MODE_PAC, MODE_MANUAL
            - Environment - MODE_ENV

            Proxy and noproxy details need to come from the same source - they are not merged
            """

            global dprint
            if debug_print is not None:
                dprint = debug_print

            self.servers = []
            self.noproxy = netaddr.IPSet([])
            self.noproxy_hosts = []

            if mode != MODE_NONE:
                # Check MODE_CONFIG and MODE_CONFIG_PAC cases
                super().__init__(mode, servers, noproxy, noproxy_hosts, debug_print)

            if self.mode == MODE_NONE:
                # Get proxy info from Internet Options
                #   MODE_AUTO, MODE_PAC or MODE_MANUAL
                ie_proxy_config = WINHTTP_CURRENT_USER_IE_PROXY_CONFIG()
                ok = ctypes.windll.winhttp.WinHttpGetIEProxyConfigForCurrentUser(
                    ctypes.byref(ie_proxy_config))
                if not ok:
                    dprint("WinHttpGetIEProxyConfigForCurrentUser failed: %s" % str(ctypes.GetLastError()))
                else:
                    if ie_proxy_config.fAutoDetect:
                        # Mode = Auto detect
                        self.mode = MODE_AUTO
                    elif ie_proxy_config.lpszAutoConfigUrl is not None and len(ie_proxy_config.lpszAutoConfigUrl) != 0:
                        # Mode = PAC
                        self.mode = MODE_PAC
                        self.servers = [ie_proxy_config.lpszAutoConfigUrl]
                        dprint("AutoConfigURL = " + self.servers[0])
                    else:
                        # Mode = Manual proxy
                        proxies = []
                        proxies_str = ie_proxy_config.lpszProxy or ""
                        for proxy_str in proxies_str.lower().replace(
                                ' ', ';').split(';'):
                            if '=' in proxy_str:
                                scheme, proxy = proxy_str.split('=', 1)
                                if scheme.strip() != "ftp":
                                    proxies.append(proxy)
                            elif proxy_str:
                                proxies.append(proxy_str)
                        if proxies:
                            self.mode = MODE_MANUAL
                            self.servers = parse_proxy(",".join(proxies))

                            # Proxy exceptions into noproxy
                            bypass_str = ie_proxy_config.lpszProxyBypass or ""
                            self.noproxy, self.noproxy_hosts = parse_noproxy(bypass_str)

            if self.mode == MODE_NONE:
                # Get from environment since nothing in Internet Options
                super().__init__()

            dprint("Proxy mode = " + str(self.mode))

        # Find proxy for specified URL using WinHttp API
        #   Used internally for MODE_AUTO, MODE_PAC and MODE_CONFIG_PAC
        def winhttp_find_proxy_for_url(self, url, autologon=True):
            ACCESS_TYPE = WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY
            if WIN_VERSION < 6.3:
                ACCESS_TYPE = WINHTTP_ACCESS_TYPE_DEFAULT_PROXY

            ctypes.windll.winhttp.WinHttpOpen.restype = ctypes.c_void_p
            hInternet = ctypes.windll.winhttp.WinHttpOpen(
                ctypes.wintypes.LPCWSTR("Px"),
                ACCESS_TYPE, WINHTTP_NO_PROXY_NAME,
                WINHTTP_NO_PROXY_BYPASS, WINHTTP_FLAG_ASYNC)
            if not hInternet:
                dprint("WinHttpOpen failed: " + str(ctypes.GetLastError()))
                return ""

            autoproxy_options = WINHTTP_AUTOPROXY_OPTIONS()
            if self.mode in [MODE_PAC, MODE_CONFIG_PAC]:
                autoproxy_options.dwFlags = WINHTTP_AUTOPROXY_CONFIG_URL
                autoproxy_options.dwAutoDetectFlags = 0
                autoproxy_options.lpszAutoConfigUrl = self.servers[0]
            elif self.mode == MODE_AUTO:
                autoproxy_options.dwFlags = WINHTTP_AUTOPROXY_AUTO_DETECT
                autoproxy_options.dwAutoDetectFlags = (
                    WINHTTP_AUTO_DETECT_TYPE_DHCP | WINHTTP_AUTO_DETECT_TYPE_DNS_A)
                autoproxy_options.lpszAutoConfigUrl = 0
            else:
                dprint("winhttp_find_proxy_for_url only applicable for MODE_AUTO, MODE_PAC and MODE_CONFIG_PAC")
                return ""
            autoproxy_options.fAutoLogonIfChallenged = autologon

            proxy_info = WINHTTP_PROXY_INFO()

            ctypes.windll.winhttp.WinHttpGetProxyForUrl.argtypes = [ctypes.c_void_p,
                ctypes.wintypes.LPCWSTR, ctypes.POINTER(WINHTTP_AUTOPROXY_OPTIONS),
                ctypes.POINTER(WINHTTP_PROXY_INFO)]
            ok = ctypes.windll.winhttp.WinHttpGetProxyForUrl(
                hInternet, ctypes.wintypes.LPCWSTR(url),
                ctypes.byref(autoproxy_options), ctypes.byref(proxy_info))
            if not ok:
                error = ctypes.GetLastError()
                if error == WINHTTP_ERROR_WINHTTP_UNABLE_TO_DOWNLOAD_SCRIPT:
                    dprint("Could not download PAC file, trying DIRECT instead")
                    return "DIRECT"
                elif error == WINHTTP_ERROR_WINHTTP_AUTODETECTION_FAILED:
                    dprint("Autodetection failed, trying DIRECT instead")
                    return "DIRECT"
                else:
                    dprint("WinHttpGetProxyForUrl failed: %s" % error)
                    return ""

            if proxy_info.dwAccessType == WINHTTP_ACCESS_TYPE_NAMED_PROXY:
                # Note: proxy_info.lpszProxyBypass makes no sense here!
                if not proxy_info.lpszProxy:
                    dprint("WinHttpGetProxyForUrl named proxy without name")
                    return ""
                return proxy_info.lpszProxy.replace(" ", ",").replace(";", ",")
            elif proxy_info.dwAccessType == WINHTTP_ACCESS_TYPE_NO_PROXY:
                return "DIRECT"
            else:
                dprint("WinHttpGetProxyForUrl accesstype %s" % (proxy_info.dwAccessType))
                return ""

            # TODO: WinHttpCloseHandle(), GlobalFree() on lpszProxy and lpszProxyBypass

        def find_proxy_for_url(self, url):
            """
            Return list of proxy servers to use for this url
              DIRECT can be returned as one of the options in the response
            """

            servers, netloc, path = super().find_proxy_for_url(url)
            if servers is not None:
                # MODE_NONE, url in no_proxy or MODE_ENV
                return servers, netloc, path
            elif self.mode in [MODE_AUTO, MODE_PAC, MODE_CONFIG_PAC]:
                # Use proxies as resolved via WinHttp
                return parse_proxy(self.winhttp_find_proxy_for_url(url)), netloc, path
            elif self.mode in [MODE_MANUAL, MODE_CONFIG]:
                # Use specific proxies configured
                return copy.deepcopy(self.servers), netloc, path
else:
    class Wproxy(_WproxyBase):
        "Load proxy information from the operating system"
        pass

if __name__ == "__main__":
    wp = Wproxy(debug_print=print)
    print("Servers: " + str(wp.servers))
    print("Noproxy: " + str(wp.noproxy))
    print("Noproxy hosts: " + str(wp.noproxy_hosts))

    wp = Wproxy(mode = MODE_CONFIG, servers = ["http://testproxy:3128"], noproxy = "google.com")
    print("Servers: " + str(wp.servers))
    print("Noproxy: " + str(wp.noproxy))
    print("Noproxy hosts: " + str(wp.noproxy_hosts))
